stumpy/scrump.py

```diff
@@ -244,10 +270,7 @@ def _compute_PI(
             continue
 
         j = nn_i
-        # Given the squared distance, work backwards and compute QT
-        QT_j = (m - P_squared[thread_idx, i, 0] / 2.0) * (Σ_T[j] * σ_Q[i]) + (
-            m * M_T[j] * μ_Q[i]
-        )
+        QT_j = QT[j]
```

Comment:
Note: Using `P_squared` to compute `QT_j` is not an alternative option anymore since the matrix profile value is post-processed internally depending on whether `T_x_subseq_isconstant` is True or False.

Reply:
So, is this a problem? I don't understand what the solution is? `QT_j = QT[j]` seems wrong?

Reply:
`QT_j = QT[j]` is okay, and there is no issue here.

I just wanted to bring your attention to this fact that `QT` of two subesquences cannot be computed based on their z-norm euclidean distance unless both of those subsequences are treated as non-constant (i.e. their corresponding value in `T_x_subseq_isconstant` is `False`). This is a minor side effect of adding param `T_x_subseq_isconstant` since it manipulates the value of distance, and hence QT cannot be computed in a backward manner. I wanted to share it with you so that if you want to use this approach (i.e. using distance to compute QT) in the future for some module, you can be careful regarding this matter.

Reply:
But how is it possible that we can set `QT_j = QT[j]`? If `QT` itself was an empty array to begin with then setting `QT_j = QT[j]` seems wrong as `QT` itself may be wrong (i.e., empty/random numbers).

Reply:
I think `QT` cannot be empty here as we use it to compute distance profile of a subsequence (based on which we find its nearest neighbor later). In other words, this is how I see it:

```
subsequence Q --> Compute QT array --> Compute distance profile for Q --> find nn of Q --> QT[at nn index]
```
In the last step, we simply get the dot product of Q and its nearest neighbor using the aleady-computed QT and the discovered nn index.  

---

Note that before we reach this part, we already have done two things:
(1) QT array is computed
(2) nn of Q is computed

And that is all we need to get the dot product between Q and its nearest neighbor, which is QT[nn].


Does this help? If your question was about something else, please let me know.

Reply:
I understand. It's just that the`QT_j = (m - P_squared[thread_idx, i, 0] / 2.0) * (Σ_T[j] * σ_Q[i]) + ( m * M_T[j] * μ_Q[i])` was taken from the original Matlab code and I can't help but feel like there was a reason for this. Simply setting `QT_j = QT[j]` feels like they they wouldn't have missed that.

Can you:

1. Create `T = np.random.rand(10000)`
2. Turn off `numba` (i.e., use pure `python` mode)
3. Keep the original code but perform a check for `if QT_j != QT[j]`
4. See if we ever enter that `if` condition
5. Repeat 10,000 times or some large number
6. Repeat for AB-joins

We need to convince ourselves that we aren't reading the code wrong

Reply:
Sure. I will update you after getting the result.

Reply:
@seanlaw 
`QT_j != QT[j]` becomes False; however, I think that is just related to loss-of-precision. The maximum absolute difference is in the order of `1e-15 `, and the test function `npt.assert_almost_equal` is passing,

Reply:
Can you please share the full test function that you used?

Reply:
As a follow up to my previous email, I checked it for AB-join, the condition `QT_j == QT[j]` is, again, not satisfied but the `npt.assert_almost_equal` is passing.

---

I also generated random time series using `np.random.uniform(-1000, 1000, n)` instead of `np.random.uniform(n)`.  Again, `QT_j` and `QT[j]` are not exactly equal (the max abs diff is in the order of `1e-09`) but they are almost equal according to `npt.assert_almost_equal`.

Reply:
Thank you. Can you double check scraamp as well to see if we need to do something similar there?